index.bs

```diff
@@ -387,233 +606,219 @@ Each \[[Stage]] value indicates the currently pending operation. If the \[[Resul
+1. If _stage_ is *undefined* then let _stageValue_ be "fetch".
+1. Else let _stageValue_ be ToString(_stage_).
+1. If _stageValue_ is an abrupt completion, return a promise rejected with a new *TypeError* exception.
+1. If IsValidStageValue(_stageValue_) is *false*, return a promise rejected with a new *RangeError* exception.
+1. Return LoadModule(_entry_, _stageValue_).
 </emu-alg>
 
-<h4 id="registry-prototype-cancel">Registry.prototype.cancel(key)</h4>
+<h4 id="module-status-prototype-result">ModuleStatus.prototype.result(stage)</h4>
 
 The following steps are taken:
 
 <emu-alg>
-1. Let _registry_ be *this* value.
-1. If Type(_registry_) is not Object, throw a *TypeError* exception.
-1. Let _entries_ be _registry_.[[RegistryData]].
-1. Let _pair_ be the entry in _entries_ such that _pair_.[[key]] is equal to _key_.
-1. If _pair_ does not exist, then throw a new *TypeError*.
-1. Let _entry_ be _pair_.[[value]].
-1. Let _stageEntry_ be GetCurrentStage(_entry_).
-1. If _stageEntry_.[[Stage]] is "link" or "ready", throw a new *TypeError*.
-1. Remove _pair_ from _entries_.
+1. Let _entry_ be *this* value.
+1. If Type(_entry_) is not Object, return a promise rejected with a new *TypeError* exception.
+1. Let _stageValue_ be ToString(_stage_).
+1. If _stageValue_ is an abrupt completion, return a promise rejected with a new *TypeError* exception.
+1. If IsValidStageValue(_stageValue_) is *false*, return a promise rejected with a new *RangeError* exception.
+1. Let _stageEntry_ be GetStage(_entry_, _stageValue_).
+1. If _stageEntry_ is an abrupt completion, return a promise rejected with a new *TypeError* exception.
+1. If _stageEntry_ is *undefined*, return a promise resolved with *undefined*.
+1. If _stageEntry_.[[Result]] is *undefined*, return a promise resolved with *undefined*.
+1. Return the result of transforming _stageEntry_.[[Result]] with a new pass-through promise.
 </emu-alg>
 
-<h4 id="registry-prototype-provide">Registry.prototype.provide(key, stage, value)</h4>
+<h4 id="module-status-prototype-provide">ModuleStatus.prototype.resolve(stage, result)</h4>
 
 The following steps are taken:
 
 <emu-alg>
-1. Let _registry_ be *this* value.
-1. If Type(_registry_) is not Object, throw a new *TypeError*.
-1. If _stage_ is not "fetch", "translate" or "instantiate", throw a new *TypeError*.
-1. Let _loader_ be _registry_.[[Loader]].
-1. Let _entry_ be EnsureRegistered(_loader_, _key_).
-1. Let _stageEntry_ be GetStage(_entry_, _stage_).
-1. If _stageEntry_ is *undefined*, throw a new *TypeError*.
-1. If _stageEntry_.[[Result]] is *undefined*, then set _entry_.[[Result]] to a promise resolved with _value_.
-1. Else, fulfill _entry_.[[Result]] with _value_.
-1. UpgradeToStage(_entry_, _stage_).
+1. Let _entry_ be *this* value.
+1. If Type(_entry_) is not Object, throw a new *TypeError*.
+1. Let _stageValue_ be ToString(_stage_).
+1. If _stageValue_ is an abrupt completion, return a promise rejected with a new *TypeError* exception.
```

Comment:
Actually _stageValue_.[[value]]. Alternately I think you can just replace this whole step with RejectIfAbrupt(stageValue).